### PR TITLE
[flutter_tools] Migrate to package:vm_service 4: trigonometric boogaloo

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description
Reland VM service migration. Fix unknown double-complete which broke flutter_test_performance.

```
test stdout (TestStep.testRunning): 00:01 +1 -1: loading /home/flutter/.cocoon/flutter/dev/automated_tests/flutter_test/trivial_widget_test.dart [E]                                                                                       
2020-04-06T10:30:42.377762: test stdout (TestStep.testRunning):   Bad state: Future already completed
2020-04-06T10:30:42.411011: test stdout (TestStep.testRunning):   dart:async/future_impl.dart 42:31                              _AsyncCompleter.complete
test stdout (TestStep.testRunning):   package:flutter_tools/src/vmservice.dart 363:29                VMService._connect.<fn>
2020-04-06T10:30:42.411101: test stdout (TestStep.testRunning):   package:stack_trace/src/stack_zone_specification.dart 209:15   StackZoneSpecification._run
test stdout (TestStep.testRunning):   package:stack_trace/src/stack_zone_specification.dart 119:48   StackZoneSpecification._registerCallback.<fn>
test stdout (TestStep.testRunning):   dart:async/zone.dart 1180:38     
```

### CODE REVIEW

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 1]
```diff
@@ -332,28 +347,60 @@ class VMService {
...
+    // Create an instance of the package:vm_service API in addition to the flutter
+    // tool's to allow gradual migration.
+    final Completer<void> streamClosedCompleter = Completer<void>();
+
+    final vm_service.VmService delegateService = vm_service.VmService(
+      primary.stream,
+      channel.add,
+      log: null,
+      disposeHandler: () async {
+        if (!streamClosedCompleter.isCompleted) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/vmservice.dart:374]: Missing check for is completed. The old VM service might have been OK with being torn down multiple times though, not sure. At any rate I can't repro the test failure anymore

Comment (zanderso) [reply packages/flutter_tools/lib/src/vmservice.dart:374]: If we're trying to tear it down twice, could we also be trying to use it after closing it in some other way?

**Review by zanderso — APPROVED** [review]
lgtm w/ question.

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/vmservice.dart:374]: I've looked into it a bit more and I'm not certain - I can't consistently repro the crash. I think the stream logic is just too complicated right now, and should be a bit simpler once I no longer need to duplicate events

**Pull request merged**